### PR TITLE
refactors AWSSigner; scheme no longer required.

### LINKED ISSUE #668: Does `AWSSigner.import_()` really require scheme string?
I imagined there would be two good ways to implement `import_()` style methods:
* either ask caller for a key identifier and then request the key details from external service
* or ask caller for all key details

Both ways have some pros and cons.... AWSSigner seems to be half-way between these: it seems to require being logged in to AWS but still requires the user to provide a scheme string.

I would expect that the scheme is in some fashion provided in the 
 `boto3_client.get_public_key()` return value, so we could improve `import_()` so that user would not actually have to provide the scheme. Does this sound correct?

**Current behavior**:

` def import_(cls, aws_key_id: str, local_scheme: str)`

**Expected behavior**:

` def import_(cls, aws_key_id: str)`
(or maybe leave local_scheme as optional for backwards compat)

cc @ianhundere

### PR DESCRIPTION
**Pull request by ianhundere**
resolves #668
### Description of the changes being introduced by the pull request:

This refactors `AWSSigner` to no longer require `local_scheme`, and instead, gets `scheme` from AWS' response (e.g. `request["SigningAlgorithms"]`).

**ecdsa**:
```
"SigningAlgorithms": [
    "ECDSA_SHA_256"
]
```

The caveat is that RSA keys have multiple `SigningAlgorithms`, so if `local_scheme` is not provided and the `keytype` is rsa then `rsassa-pss-sha256` is chosen by default as is the case throughout `securesystemslib`.

**rsa**:
```
"SigningAlgorithms": [
    "RSASSA_PKCS1_V1_5_SHA_256",
    "RSASSA_PKCS1_V1_5_SHA_384",
    "RSASSA_PKCS1_V1_5_SHA_512",
    "RSASSA_PSS_SHA_256",
    "RSASSA_PSS_SHA_384",
    "RSASSA_PSS_SHA_512"
]
```

Also includes updated docs and error handling.

### Please verify and check that the pull request fulfills the following requirements:

- [x] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [x] Tests have been added for the bug fix or new feature
- [x] Docs have been added for the bug fix or new feature

cc @jku @lukpueh

### CODE REVIEW

File: securesystemslib/signer/_aws_signer.py
Comment (ianhundere) [reply securesystemslib/signer/_aws_signer.py]: ah, thanks for catching that. i definitely was overthinking it. 😅

File: securesystemslib/signer/_aws_signer.py
Comment (jku) [inline securesystemslib/signer/_aws_signer.py:233]: Good addition, thanks: I had not realized this is not prehashed.

This is not for this PR but just as a note:
* Users probably expect prehashing (instead of sending potentially large files to AWS)
* Confusingly this method does prehash the content... the result is just not used

**Review by jku — APPROVED** [review]
Thanks. I think this is good to merge: it improves `import_()` in that it can be used without the other argument.


I'll leave some notes below but I am **not** requesting changes: I have no access to AWS signer so I can't really tell what is smart and what is not.

So just as a comment: the lookup (from-AWS-response-to-securesystemslib-key-type-and-scheme) still looks more complicated than it probably is 
* RSA users can now choose the hash length but ECDSA users for some reason cannot?
* the real import_() argument probably should have been `hash_size: int = 256` if that's the only real variable. This would be an API change for AWSSigner
* then the single lookup function  would be `def _get_keytype_and_scheme(aws_response, hash_size) -> Tuple[str, str]`

**Review by lukpueh — CHANGES_REQUESTED** [review]
Sorry for commenting late. I am not a big fan of the new key-type specific argument. It breaks the API and feels inconsistent.

I suggest to keep the `local_scheme`, make it optional, and use it like so:
```
get key from aws by keyid

if scheme is passed:
  if scheme is not allowed for key
    raise

else:
  assign default scheme for key
```

I don't know exactly what schemes are supported for what key, but I expect that for any rsa key in the kms you should be able to use any of the securesystemslib schemes, whereas for ecdsa there is only one scheme per key.

A very brief look at the [docs](https://docs.aws.amazon.com/kms/latest/developerguide/asymmetric-key-specs.html#key-spec-rsa) seems to support my assumption.

**Comment by ianhundere** [comment]
@jku @lukpueh thanks for y'all's input, i've gone ahead and renamed `rsa-scheme` back to `local_scheme`.

File: securesystemslib/signer/_aws_signer.py
Comment (lukpueh) [inline securesystemslib/signer/_aws_signer.py:121]: It looks like `local_scheme` is ignored now, if the key type is ecdsa. I think we should fail, if the user passes an unsupported scheme.

File: securesystemslib/signer/_aws_signer.py
Comment (lukpueh) [reply securesystemslib/signer/_aws_signer.py:233]: > Confusingly this method does prehash the content... the result is just not used

If you are not sending the prehashed content, I suggest to  remove the code that hashes the content. This means you no longer need `self.hash_algorithm`, which means that you no longer need `self._get_aws_signing_algo` (except maybe for catching unsupported schemes).

File: securesystemslib/signer/_aws_signer.py
Comment (lukpueh) [inline securesystemslib/signer/_aws_signer.py:123]: Same here. If `local_scheme` is unsupported for the rsa key you won't fail before calling `_get_aws_signing_algo` for the constructor, which is actually an unnecessary operation (see comment above in `sign`).

What do you think about the following strategy, if `local_scheme` is passed:
1. get aws representation (e.g. RSASSA_PSS_SHA_256) for passed securesystemslib-style scheme (e.g. rsassa-pss-sha256)
2. fail if there it doesn't translate 
3. fail if the scheme is not in the returned SigningAlgorithms

**Review by lukpueh — CHANGES_REQUESTED** [review]

**Comment by lukpueh** [comment]
fyi @ianhundere, I've tested AWSSigner with Localstack for the first time today and it works like a charm! 🚀 ❤️ 
https://github.com/repository-service-tuf/repository-service-tuf-worker/pull/452

Thanks for your efforts!

File: securesystemslib/signer/_aws_signer.py
Comment (ianhundere) [reply securesystemslib/signer/_aws_signer.py:123]: sounds good, i'm on holiday but will implement your feedback when i'm back.

File: securesystemslib/signer/_aws_signer.py
Comment (lukpueh) [inline securesystemslib/signer/_aws_signer.py:119]: This is the wrong scheme to be used for SSlibKey below.

I ran my [localstack test from RSTUF](https://github.com/repository-service-tuf/repository-service-tuf-worker/pull/452) against your PR and it failed with this error message:

```
.tox/local-aws-kms/lib/python3.10/site-packages/securesystemslib/signer/_aws_signer.py:67: in __init__
    self.aws_algo = self._get_aws_signing_algo(self.public_key.scheme)
...
E       KeyError: 'RSASSA_PSS_SHA_256'
```

File: securesystemslib/signer/_aws_signer.py
Comment (lukpueh) [inline securesystemslib/signer/_aws_signer.py:121]: nit: since you raise here without having looked at the fetched key, you could consider doing the check before fetching the key.

File: securesystemslib/signer/_aws_signer.py
Comment (lukpueh) [inline securesystemslib/signer/_aws_signer.py:129]: Same as above, you could check for this case too, if the fetched key actually supports the default scheme.

**Review by lukpueh — CHANGES_REQUESTED** [review]
Thanks for the updates! There is one small bug in `import_` which needs to be fixed before we can merge this. The rest is polishing.

File: securesystemslib/signer/_aws_signer.py
Comment (lukpueh) [reply securesystemslib/signer/_aws_signer.py]: It doesn't look like you fixed this bug. When I pass `"rsassa-pss-sha256"` to `import_`, it returns an `SSlibKey` with  `"RSASSA_PSS_SHA_256"` as `scheme`. Later, when I pass that key to  `AWSigner.__init__`, I get the above error.

File: securesystemslib/signer/_aws_signer.py
Comment (lukpueh) [reply securesystemslib/signer/_aws_signer.py]: >  you might also want to check, if the fetched key actually supports the passed scheme

Did you decide to not check it? I'm fine with that, maybe we can write an issue.

File: securesystemslib/signer/_aws_signer.py
Comment (lukpueh) [reply securesystemslib/signer/_aws_signer.py]: see https://github.com/secure-systems-lab/securesystemslib/pull/724#discussion_r1526029353

File: securesystemslib/signer/_aws_signer.py
Comment (ianhundere) [reply securesystemslib/signer/_aws_signer.py]: doh, rushed it yesterday.

File: securesystemslib/signer/_aws_signer.py
Comment (ianhundere) [reply securesystemslib/signer/_aws_signer.py]: ah, thanks for the heads up. i think that's fixed now.

File: securesystemslib/signer/_aws_signer.py
Comment (lukpueh) [inline securesystemslib/signer/_aws_signer.py:129]: You are mixing EAFP and LBYL here with the consequence that your ValueError will never be raised. :) 

I suggest to write:

```suggestion
        if local_scheme:
            if local_scheme not in cls.aws_signing_algorithms:
                raise ValueError(f"Unsupported scheme: {local_scheme}") from e
                
        # no need to rename `local_scheme`, you can just use this name below
```

**Review by lukpueh — COMMENTED** [review]
Thanks for your persistence! It looks like you are still not consistently checking, if the key supports the chosen scheme according to `SigningAlgorithms` in the response.

As I said, I'm okay, if we improve that in a follow-up PR. 

Please address my inline comment and we can merge this (or let me know, if I should address it).

File: securesystemslib/signer/_aws_signer.py
Comment (ianhundere) [reply securesystemslib/signer/_aws_signer.py]: ah, thank you.

Comment (lukpueh) [reply securesystemslib/signer/_aws_signer.py]: Very welcome!

Sorry if my comment about `local_scheme` was unclear. If you don't rename it here, you will need to use `local_scheme` below. See my code suggestions...

File: securesystemslib/signer/_aws_signer.py
Comment (lukpueh) [inline securesystemslib/signer/_aws_signer.py:151]:
```suggestion
                local_scheme = "rsassa-pss-sha256"
```

File: securesystemslib/signer/_aws_signer.py
Comment (lukpueh) [inline securesystemslib/signer/_aws_signer.py:162]:
```suggestion
        keyid = compute_default_keyid(keytype, local_scheme, keyval)
        public_key = SSlibKey(keyid, keytype, local_scheme, keyval)
```

File: securesystemslib/signer/_aws_signer.py
Comment (lukpueh) [inline securesystemslib/signer/_aws_signer.py:130]: Oh and the try/except block should also go away.

**Comment by ianhundere** [comment]
> Thanks for the updates! There is one small bug in `import_` which needs to be fixed before we can merge this. The rest is polishing.

okay, i think that about does it; thanks for all your feedback. 🙇

File: securesystemslib/signer/_aws_signer.py
Comment (lukpueh) [inline securesystemslib/signer/_aws_signer.py:156]: Don't these checks always raise when `local_scheme` is not passed?

Comment (ianhundere) [reply securesystemslib/signer/_aws_signer.py:156]: ah, you're correct. don't think these checks are necessary
```
        if (
            keytype == "ecdsa"
            and cls._get_ecdsa_scheme(aws_scheme) != local_scheme
        ):
            raise ValueError(
                f"The AWS KMS key does not support the scheme: {local_scheme}"
            )
        if (
            keytype == "rsa"
            and local_scheme not in cls.aws_signing_algorithms.values()
        ):
            raise ValueError(
                f"The AWS KMS key does not support the scheme: {local_scheme}"
            )
```

Comment (lukpueh) [reply securesystemslib/signer/_aws_signer.py:156]: Let's add something like this in a follow-up PR.

Comment (ianhundere) [reply securesystemslib/signer/_aws_signer.py:163]: ah, i just pushed new commit.

Comment (lukpueh) [reply securesystemslib/signer/_aws_signer.py:163]: no need to revert. I think, now that you swapped the blocks, it works. let me give it one final look.

File: securesystemslib/signer/_aws_signer.py
Comment (lukpueh) [inline securesystemslib/signer/_aws_signer.py:159]: 🙈 this compares a string like "rsassa-pss-sha256"  to strings like "RSASSA_PSS_SHA_256"...

File: securesystemslib/signer/_aws_signer.py
Comment (lukpueh) [inline securesystemslib/signer/_aws_signer.py:163]: Let's remove these checks after all. And add them after #612...

```suggestion
```

**Review by lukpueh — CHANGES_REQUESTED** [review]

File: securesystemslib/signer/_aws_signer.py
Comment (ianhundere) [reply securesystemslib/signer/_aws_signer.py]: oops, should have been:
```
        if keytype == "ecdsa" and local_scheme != cls._get_ecdsa_scheme(
            aws_scheme
        ):
            raise ValueError(
                f"The AWS KMS key does not support the scheme: {local_scheme}"
            )

        if (
            keytype == "rsa"
            and local_scheme not in cls.aws_signing_algorithms
        ):
            raise ValueError(
                f"The AWS KMS key does not support the scheme: {local_scheme}"
            )
```

File: securesystemslib/signer/_aws_signer.py
Comment (ianhundere) [reply securesystemslib/signer/_aws_signer.py]: ah okay, i can i add what i mentioned above after https://github.com/secure-systems-lab/securesystemslib/issues/612.

**Review by lukpueh — APPROVED** [review]

**Pull request merged**
